src/Microsoft.AspNetCore.Server.Kestrel/Http/Frame.cs

```diff
@@ -174,16 +174,25 @@ public bool HasResponseStarted
 
         public void InitializeHeaders()
         {
-            _frameHeaders = HttpComponentFactory.CreateHeaders(DateHeaderValueManager);
-            RequestHeaders = _frameHeaders.RequestHeaders;
-            ResponseHeaders = _frameHeaders.ResponseHeaders;
+            if (_frameHeaders == null)
+            {
+                _frameHeaders = new Headers();
+                RequestHeaders = _frameHeaders.RequestHeaders;
+                ResponseHeaders = _frameHeaders.ResponseHeaders;
+            }
+
+            _frameHeaders.Initialize(DateHeaderValueManager);
         }
 
 
         public void InitializeStreams(MessageBody messageBody)
         {
-            _frameStreams = HttpComponentFactory.CreateStreams(this);
+            if (_frameStreams == null)
+            {
+                _frameStreams = new Streams();
+            }
 
+            _frameStreams.Initialize(this);
```

Comment:
Initialize doesn't do anything for streams that couldn't be done in the constructor.